web/src/main/java/org/springframework/security/web/csrf/CsrfFilter.java

```diff
@@ -128,7 +121,8 @@ protected void doFilterInternal(HttpServletRequest request, HttpServletResponse
 		CsrfToken csrfToken = deferredCsrfToken.get();
 		String actualToken = this.requestHandler.resolveCsrfTokenValue(request, csrfToken);
 		if (!equalsConstantTime(csrfToken.getToken(), actualToken)) {
-			boolean missingToken = deferredCsrfToken.isGenerated();
+			boolean missingToken = DeferredCsrfToken.class.isAssignableFrom(deferredCsrfToken.getClass())
```

Comment:
This is implementation specific as there is nothing in the interface that says that the `Supplier<CsrfToken>` will return a `DeferredCsrfToken` if it is generated.

With `DeferredCsrfToken` package private, it would be difficult for someone to properly implement a custom `CsrfTokenRepository` and overrride the `loadToken(request, response)` method since they cannot use `DeferredCsrfToken`.

I wonder if we should change the return type of the repository to `DeferredCsrfToken`

Reply:
I'd prefer to add `DeferredCsrfToken`. It differs a bit from `SecurityContextRepository` in that we need to know if the `CsrfToken` was generated or loaded and as such we should have an explicit API. The `GENERATED_CSRF_TOKEN_ATTR_NAME` is a new contract (much like having a new interface) but is more difficult for users to use/discover.